### PR TITLE
Closes #36. Newtonsoft API replaced with Thoth

### LINKED ISSUE #36: Use Thoth.Json.Net everywhere consistently
Part of [this review](https://github.com/Zaid-Ajaj/Femto/pull/35#discussion_r297655462) by @MangelMaxime 

Use `Thoth.Json.Net` instead of the low-level `Newtonsoft.Json`

### PR DESCRIPTION
**Pull request by semuserable**
Attempt to replace Newtonsoft API with Thoth.

### CODE REVIEW

File: Program.fs [hunk 1]
```diff
@@ -37,6 +36,11 @@ type ResolveAction =
     | UninstallDev of library:string * package: string * version:string
     | UnableToResolve of library:string * package:string * range:string * error: string
 
+type RawPackage = {
```
Comment (Zaid-Ajaj) [inline Program.fs]: Suggestion: rename `RawPackage` to `PackageJson` because these fields are read from a package.json file

File: Program.fs [hunk 2]
```diff
@@ -80,37 +84,25 @@ let needsNodeModules (packageJson: string) =
...
+                    Range = Some (SemVer.Range(pair.Value));
+                    Installed = None;
+                    DevDependency = isDevDependency
+                })
+            
+            ResizeArray [
+                yield! createInstalledPackages false rawPackage.Dependencies;
+                yield! createInstalledPackages true rawPackage.DevDependencies
+            ]
+        | Error _ -> ResizeArray []
```
Comment (Zaid-Ajaj) [inline Program.fs]: Suggestion: log the parsing error to the console before returning the empty `ResizeArray [ ]`

File: Program.fs [hunk 3]
```diff
@@ -130,10 +122,18 @@ let findInstalledPackages (packageJson: string) : ResizeArray<InstalledNpmPackag
             if not (IO.File.Exists pkgJson)
                 then ()
             else
-                let pkgJsonContent = JObject.Parse(File.readAllTextNonBlocking pkgJson)
+                let nameAndVersionDecoder = Decode.object (fun it ->
```
Comment (Zaid-Ajaj) [inline Program.fs]: Suggestion: use `get` instead of `it` as per @MangelMaxime recommendation because this is a "field getter"

File: Program.fs [hunk 4]
```diff
@@ -130,10 +122,18 @@ let findInstalledPackages (packageJson: string) : ResizeArray<InstalledNpmPackag
...
+                    (it.Required.Field "name" Decode.string,
+                     it.Required.Field "version" Decode.string)
+                )
+                
+                let (name, version) =
+                    File.readAllTextNonBlocking pkgJson
+                    |> Decode.unsafeFromString nameAndVersionDecoder
```
Comment (Zaid-Ajaj) [inline Program.fs]: Please avoid unsafe decoding and log the parsing errors to the console if any

File: Program.fs [hunk 5]
```diff
@@ -952,12 +950,20 @@ let installPaketFromBootstrapper projectRoot =
...
+                )
+                |> Decode.list
+                |> Decode.field "data"
+                
             let latestPaketVersion =
-                packages
-                |> Seq.tryFind (fun pkg -> pkg.["id"].ToObject<string>() = "Paket")
-                |> Option.map (fun pkg -> pkg.["version"].ToObject<string>())
+                paketSearchResult
+                |> Decode.unsafeFromString dataDecoder
```
Comment (Zaid-Ajaj) [inline Program.fs]: Just like the above comment, avoid unsafe decoding please

**Comment by semuserable** [comment]
Thank you for the review @Zaid-Ajaj!

Regarding `unsafe` calls, I just noticed that there were `unsafe` calls already and decided to accommodate existing logic with it, so if I see that there is a direct call without checks, then I do `unsafe` version :)

I agree with all the comments and fix them as soon as I have time :)

**Comment by semuserable** [comment]
New changes!

@Zaid-Ajaj please check, I hope I haven't missed anything :)

**Comment by Zaid-Ajaj** [comment]
Awesome stuff @semuserable :fire: I will publish a new version soon. Thanks again for the PR 🙏

**Comment by semuserable** [comment]
That's so cool @Zaid-Ajaj! Thank you!

**Pull request merged**